bridges/snowbridge/primitives/router/src/inbound/v2.rs

```diff
@@ -0,0 +1,642 @@
+	EthereumUniversalLocation,
+	GlobalAssetHubLocation,
+> where
+	EthereumNetwork: Get<NetworkId>,
+	InboundQueuePalletInstance: Get<u8>,
+	ConvertAssetId: MaybeEquivalence<TokenId, Location>,
+	GatewayProxyAddress: Get<H160>,
+	EthereumUniversalLocation: Get<InteriorLocation>,
+	GlobalAssetHubLocation: Get<Location>,
+{
+	_phantom: PhantomData<(
+		EthereumNetwork,
+		InboundQueuePalletInstance,
+		ConvertAssetId,
+		GatewayProxyAddress,
+		EthereumUniversalLocation,
+		GlobalAssetHubLocation,
+	)>,
+}
+
+impl<
+		EthereumNetwork,
+		InboundQueuePalletInstance,
+		ConvertAssetId,
+		GatewayProxyAddress,
+		EthereumUniversalLocation,
+		GlobalAssetHubLocation,
+	> ConvertMessage
+	for MessageToXcm<
+		EthereumNetwork,
+		InboundQueuePalletInstance,
+		ConvertAssetId,
+		GatewayProxyAddress,
+		EthereumUniversalLocation,
+		GlobalAssetHubLocation,
+	>
+where
+	EthereumNetwork: Get<NetworkId>,
+	InboundQueuePalletInstance: Get<u8>,
+	ConvertAssetId: MaybeEquivalence<TokenId, Location>,
+	GatewayProxyAddress: Get<H160>,
+	EthereumUniversalLocation: Get<InteriorLocation>,
+	GlobalAssetHubLocation: Get<Location>,
+{
+	fn convert(
+		message: Message,
+		origin_account_location: Location,
+	) -> Result<(Xcm<()>, u128), ConvertMessageError> {
+		let mut message_xcm: Xcm<()> = Xcm::new();
+		if message.xcm.len() > 0 {
+			let xcm_string = hex::encode(message.xcm.clone());
+			log::info!(target: LOG_TARGET,"found xcm payload: {:x?}", xcm_string);
+			// Allow xcm decode failure so that assets can be trapped on AH instead of this
+			// message failing but funds are already locked on Ethereum.
+			if let Ok(versioned_xcm) = VersionedXcm::<()>::decode_with_depth_limit(
+				MAX_XCM_DECODE_DEPTH,
+				&mut message.xcm.as_ref(),
+			) {
+				if let Ok(decoded_xcm) = versioned_xcm.try_into() {
+					message_xcm = decoded_xcm;
+				} else {
+					log::error!(target: LOG_TARGET,"unable to decode xcm");
+				}
+			} else {
+				log::error!(target: LOG_TARGET,"unable to decode versioned xcm");
+			}
+		}
+
+		log::info!(target: LOG_TARGET,"xcm decoded as {:?}", message_xcm);
+
+		let network = EthereumNetwork::get();
+
+		// use eth as asset
+		let fee_asset = Location::new(2, [GlobalConsensus(EthereumNetwork::get())]);
+		let fee: XcmAsset = (fee_asset.clone(), message.execution_fee).into();
+		let eth: XcmAsset =
+			(fee_asset.clone(), message.execution_fee.saturating_add(message.value)).into();
+		let mut instructions = vec![
+			DescendOrigin(PalletInstance(InboundQueuePalletInstance::get()).into()),
+			UniversalOrigin(GlobalConsensus(network)),
+			ReserveAssetDeposited(eth.into()),
+			PayFees { asset: fee },
+		];
+		let mut reserve_assets = vec![];
+		let mut withdraw_assets = vec![];
+
+		let mut refund_surplus_to = origin_account_location;
+
+		if let Some(claimer) = message.claimer {
+			// If the claimer can be decoded, add it to the message. If the claimer decoding fails,
+			// do not add it to the message, because it will cause the xcm to fail.
```

Comment:
yes, not clear if we should fail the whole thing or just continue with default claimer...

Reply:
I have an idea of a more general solution for any type of failures on inbound queue that are caused by protocol incompatibilities (bridge's fault, not the user's):

the inbound queue verifies the authenticity and exact validity of the message through the `prover`, so if it fails on some decoding step we can consider the message invalid, it will always fail, and hopefully we can revert it the Ethereum side.
So maybe we can mark it in some other `InvalidMessage: SparseBitmap` storage item or simply emit an `InvalidMessage(nonce)` event that can be proved back to Ethereum so that it can be rolled back there in case of these kinds of bridge protocol issues.

This is a further enhancement idea to be done outside of this PR, but worth exploring.

Reply:
This makes sense :+1:

Reply:
Same comment as here: https://github.com/paritytech/polkadot-sdk/pull/6697#discussion_r1940757212 It might be better to allow the invalid claimer, so that the funds can be trapped on AH.

The future improvement sounds cool!

Reply:
If the claimer is invalid, we set the Snowbridge sovereign account to the claimer: https://github.com/vgeddes/polkadot-sdk/blob/snowbridge-v2/bridges/snowbridge/primitives/inbound-queue/src/v2/converter.rs#L120-L123
https://github.com/vgeddes/polkadot-sdk/blob/snowbridge-v2/bridges/snowbridge/primitives/inbound-queue/src/v2/converter.rs#L311C7-L321